### PR TITLE
Improve listing performance of Hudi tables

### PR DESCRIPTION
**Pull request by codope**
- Integrate metadata-based listing for Hudi tables. This is
enabled by a session property.
- Implement a new DirectoryLister for Hudi that uses
HoodieTableFileSystemView to fetch data files.
- Bump Hudi version to 0.10.0 to use above features.

Test plan - 
Synced a Hudi table to Hive and queried through Presto
with both metadata enabled and disabled.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: pom.xml
Comment (vinothchandar) [inline pom.xml:72]: this needs to be called out in the PR summary?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java
Comment (vinothchandar) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:1732]: lets make this description more future proof. 

"For Hudi tables prefer to fetch the list of file names, sizes and other metadata from the internal metadata table rather than storage"

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (vinothchandar) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:409]: is n't this check same as `!isHudiParquetInputFormat(inputFormat)` above? any chances to simplify?

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java
Comment (vinothchandar) [inline presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:110]: can we just call the same API here? `getLatestBaseFiles()` and simplify? i think `getLatestBaseFiles()` calls `fetchLatestBaseFiles`

**Review by vinothchandar — COMMENTED** [review]
Minor comments. Looking in good shape!

File: presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java:263]: Yes. I kept it to be compatible in case users disable and not use `HudiDirectoryLister`. But, actually we introduced the path filter in #13818 . Let's remove it in a follow up PR.

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:110]: `getLatestBaseFiles()` also calls `ensurePartitionLoadedCorrectly()` which adds files to fs view. Since, in this case we have already done that, so directly calling `fetchLatestBaseFiles ()`.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:409]: `HoodieParquetRealtimeInputFormat` is subclass of `HoodieParquetInputFormat`. However, we want to use file splits from the former but not latter. Hence, a separate check.

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java
Comment (arunthirupathi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:69]: Is this method an override ? If not, why does it return optional, though the underlying value is always present.

**Review by arunthirupathi — COMMENTED** [review]
Can this PR contain some unit test instead of manual verifiication ?

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java
Comment (codope) [reply presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:69]: Yeah, optional is not needed here. I will remove it.

**Comment by codope** [comment]
> Can this PR contain some unit test instead of manual verifiication ?

Hi @arunthirupathi thanks for reviewing the PR. I have added a test for `HudiDirectoryLister`. Can you take a look again?

File: pom.xml
Comment (arunthirupathi) [inline pom.xml:1210]: Why is hbase dependency required ?

File: presto-hive/pom.xml
Comment (arunthirupathi) [inline presto-hive/pom.xml:61]: Hbase is a pretty big dependency and this could cause issues.

Comment (codope) [reply presto-hive/pom.xml:61]: Hudi uses HBase in two ways:
1. The metadata files are written as HFile.
2. To maintain index of Hudi filegroup id to external bootstrap file.
In this PR, we just need `Bytes` class from hbase-common for initializing `HoodieTableFileSystemView` and I have excluded deps that we don't need. We have created a thinner [hudi-presto-bundle](https://github.com/apache/hudi/pull/3623) which I plan to update in presto-hive  pom in a subsequent PR.

Comment (arunthirupathi) [reply presto-hive/pom.xml:61]: Currently Presto does not pull in other bigger open source jars (hadoop, hive) and the only way to do is to fork the code reduce it to only bare minimal code and then take a dependency on it. 

https://github.com/prestodb/presto-hadoop-apache2
https://github.com/prestodb/presto-hive-apache

Hbase will cause dependency hell for Presto as some of the dependencies are conflicting. https://mvnrepository.com/artifact/org.apache.hbase/hbase-common/2.4.8

This will be a blocker to merge this PR.

Comment (arunthirupathi) [reply presto-hive/pom.xml:61]: Even linked [PR](https://github.com/apache/hudi/pull/3623/files), shades everything which will avoid dependency issues but it will cause the jar size to explode. 

What is the size of the shaded jar ?

Comment (codope) [reply presto-hive/pom.xml:61]: It's about [13 MB](https://mvnrepository.com/artifact/org.apache.hudi/hudi-presto-bundle/0.10.0). 
Unfotunately, HFile is only packaged as part of HBase. Would it be more preferable to use hudi-presto-bundle instead of adding hbase directly?

Comment (codope) [reply presto-hive/pom.xml]: @arunthirupathi Can you please review this again? Check the latest commit where i'm using the stripped down hudi-presto-bundle.

Comment (arunthirupathi) [reply presto-hive/pom.xml]: For presto-hive can you please do a dependency tree diff of before and after and paste what changed ?

mvn dependency:tree -pl presto-hive (run this on master and the change branch)

Comment (codope) [reply presto-hive/pom.xml]: Below is the diff. The full dependency tree is in [this gist](https://gist.github.com/codope/cb0d174d371e8be698d0342f5cf3393f).
```
sagars@Sagars-MacBook-Pro presto % git diff --no-index presto_hive_master_dep.txt presto_hive_hudi_dep.txt
diff --git a/presto_hive_master_dep.txt b/presto_hive_hudi_dep.txt
index 3f6e69f26a..21733b24f4 100644
--- a/presto_hive_master_dep.txt
+++ b/presto_hive_hudi_dep.txt
@@ -20,8 +20,7 @@
 [INFO] +- com.facebook.presto:presto-expressions:jar:0.268-SNAPSHOT:compile
 [INFO] +- com.facebook.presto:presto-cache:jar:0.268-SNAPSHOT:compile
 [INFO] |  \- org.alluxio:alluxio-shaded-client:jar:2.7.0:compile
-[INFO] +- org.apache.hudi:hudi-common:jar:0.9.0:compile
-[INFO] +- org.apache.hudi:hudi-hadoop-mr:jar:0.9.0:compile
+[INFO] +- org.apache.hudi:hudi-presto-bundle:jar:0.10.0:compile
 [INFO] +- com.facebook.presto:presto-memory-context:jar:0.268-SNAPSHOT:compile
 [INFO] +- com.facebook.presto:presto-hive-common:jar:0.268-SNAPSHOT:compile
 [INFO] +- com.facebook.presto:presto-hive-metastore:jar:0.268-SNAPSHOT:compile
@@ -249,6 +248,6 @@
 [INFO] ------------------------------------------------------------------------
 [INFO] BUILD SUCCESS
 [INFO] ------------------------------------------------------------------------
-[INFO] Total time:  2.809 s
-[INFO] Finished at: 2021-12-23T12:51:07+05:30
+[INFO] Total time:  2.857 s
+[INFO] Finished at: 2021-12-23T12:52:44+05:30
 [INFO] ------------------------------------------------------------------------
```

File: presto-iceberg/pom.xml
Comment (arunthirupathi) [inline presto-iceberg/pom.xml:427]: Why is the exclusion required here and other directories ?

**Review by arunthirupathi — APPROVED** [review]
left one comment, but the rest looks good.

File: presto-iceberg/pom.xml
Comment (codope) [reply presto-iceberg/pom.xml:427]: Without ignoring, the maven duplicate finder plugin complains about "Found duplicate (but equal) classes in [module A, module B]."

Comment (arunthirupathi) [reply presto-iceberg/pom.xml:427]: Does the prersto-hudi bundle contains copy of this class , but not shaded ?

**Review by arunthirupathi — COMMENTED** [review]
There are failing tests, Can you please rebase and squash the commits ?

File: presto-iceberg/pom.xml
Comment (codope) [reply presto-iceberg/pom.xml:427]: Yeah. I just created a PR to shade them https://github.com/apache/hudi/pull/4495

**Comment by codope** [comment]
> There are failing tests, Can you please rebase and squash the commits ?

Done. But looks like there is one flaky test - `TestDistributedQueryResource#testGetAllQueryInfoForLimits`. It passes locally.

File: presto-iceberg/pom.xml
Comment (arunthirupathi) [reply presto-iceberg/pom.xml:427]: When will the new presto-hudi bundle available ? Can you use the new presto-hudi bundle instead of this one ?

**Comment by arunthirupathi** [comment]
For the failing tests, rebase and push couple of times. If it fails after two attempts let me know, I can take a look and either disable or reach out to the author and fix it.

File: presto-iceberg/pom.xml
Comment (vinothchandar) [reply presto-iceberg/pom.xml:427]: @arunthirupathi We plan a 0.10.1 minor release in Jan. Can we merge this and we can follow up after? Do you see any issues with this exclusion.

Comment (arunthirupathi) [reply presto-iceberg/pom.xml:427]: > @arunthirupathi We plan a 0.10.1 minor release in Jan. Can we merge this and we can follow up after? Do you see any issues with this exclusion.

I don't see any, if you folks make effort to release this in Jan and remove the exclusion later. As usual, I will revert on first sign of trouble.

Comment (arunthirupathi) [reply presto-iceberg/pom.xml:427]: Once the test passes,  I will merge this change.

**Comment by cocozianu** [comment]
We've been running into trouble with his approach and in particular with hudi-presto-bundle 0.10.0
I created the issue: https://github.com/prestodb/presto/issues/17164
and the associated revert pull request
https://github.com/prestodb/presto/pull/17165

**Comment by arunthirupathi** [comment]
First, I apologize for the issue caused. I was concerned about the unshaded classes, but all the tests passed, so I thought this was not an issue.

Can you please share more details about the errors you are running into ? I will take a look and see if I can add more tests/errors to exercise them. Give as much details as possible, to prevent future regressions.

**Comment by cocozianu** [comment]
Hi Sagar, Sumit.

Current conflicts, in facebook's installation of presto:

Found duplicate and different classes in [com.facebook.presto.hive:hive-apache:3.0.0-7, org.apache.hudi:hudi-presto-bundle:0.10.0]:
[WARNING]   org.apache.parquet.avro.AvroCompatRecordMaterializer
[WARNING]   org.apache.parquet.avro.AvroDataSupplier
[WARNING]   org.apache.parquet.avro.AvroIndexedRecordConverter
[WARNING]   org.apache.parquet.avro.AvroParquetInputFormat
[WARNING]   org.apache.parquet.avro.AvroParquetOutputFormat
[WARNING]   org.apache.parquet.avro.AvroParquetWriter
[WARNING]   org.apache.parquet.avro.AvroReadSupport
[WARNING]   org.apache.parquet.avro.AvroRecordConverter
[WARNING]   org.apache.parquet.avro.AvroRecordMaterializer
[WARNING]   org.apache.parquet.avro.AvroSchemaConverter
[WARNING]   org.apache.parquet.avro.AvroWriteSupport
[WARNING]   org.apache.parquet.avro.GenericDataSupplier
[WARNING]   org.apache.parquet.avro.ParentValueContainer
[WARNING]   org.apache.parquet.avro.ReflectDataSupplier
[WARNING]   org.apache.parquet.avro.SpecificDataSupplier

**Comment by vinothchandar** [comment]
Hmmm. Not sure if we can shade parquet-avro, but we could skip bundling it with the hudi-presto-bundle. We will take a closer look at the hudi-presto-bundle, we have a 0.10.1 upcoming. 

Apologies for your trouble!

**Pull request merged**
